### DISCUSSION TITLE
Proper way to do type checking

### DISCUSSION THREAD

**Discussion by ymc9** [eclipse-langium/langium#706] [Q&A] [answered]
Hi guys,

I'm building a DSL and making good progress with Langium. Thank you for the great work!

My DSL has a complex expression system that needs to be type-checked (resolving expression types). Just wondering what's the proper way to do that:

1. I believe I should do type-checking at linking time. In the same pass as linking or maybe another pass after that. Just wondering if it's appropriate.
2. How can I extend the generated AstNode types to carry typing information? Right now, I'm just adding an extra field by casting "any" back and forth. Is there a good way to achieve this?
3. Is there an existing example of type-checking?

Thank you!

**Comment by msujew** [comment, marked as answer]
Hey @ymc9,

> I believe I should do type-checking at linking time. In the same pass as linking or maybe another pass after that. Just wondering if it's appropriate.

Yeah, we do the same. It kind of depends on what your type system looks like, but I assume it contains something like interfaces/classes/structured types. In that case the linking step already requires a working type system, as you need it to resolve to fields/methods of classes for example.

> How can I extend the generated AstNode types to carry typing information? Right now, I'm just adding an extra field by casting "any" back and forth. Is there a good way to achieve this?

We usually just store a map for `AstNode` -> `type` relationships on every document, which is effectively the same.

Note that you need to see whether that also works correctly in case the document is part of an affected document rebuild cycle (i.e. a linked document has changed and the document in question needs to be rebuilt as well). We don't reparse the document in that case, and just throw away all references in that document. Depending on how you set this up, some of your types may become stale due to this.

> Is there an existing example of type-checking?

Unfortunately, not yet. We have some stuff internally, but it's under NDAs. We're currently thinking about building a type-checking framework, which would alleviate most of the heavy-lifting one needs to do regarding type checking, but there's nothing concrete in this direction yet. Probably something that we would tackle in Q1 of 2023.

**Reply by ymc9** [reply]
Thank you! Great to know that I'm on the right track 😄. Yes, my DSL involves resolving field references to declarations, so I'm right now mixing reference resolution and type checking together. The default implementation of DefaultLinker is helpful.

Looking forward to trying out the type-checking framework!

**Reply by msujew** [reply]
Small update: We actually now have a publicly available example for type checking, available over at the [`langium-lox` repo](https://github.com/langium/langium-lox/tree/type-system). Make sure to select the `type-system` branch, as that part is still work in progress and will likely take a few days until it is merged.

**Reply by ymc9** [reply]
Cool! I'll look into the sample. Thank you for the great work!
